samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,571 @@
+
+		if (ret < 0) {
+			int err;
+
+			if (ret != -EAGAIN) {
+				return ret;
+			}
+			printk("mqtt_read_publish_payload: EAGAIN\n");
+
+			err = poll(&fds, 1, K_SECONDS(CONFIG_MQTT_KEEPALIVE));
+			if (err > 0 && (fds.revents & POLLIN) == POLLIN) {
+				continue;
+			} else {
+				return -EIO;
+			}
+		}
+
+		if (ret == 0) {
+			return -EIO;
+		}
+
+		buf += ret;
+	}
+
+	return 0;
+}
+
+/**@brief MQTT client event handler
+*/
+void mqtt_evt_handler(struct mqtt_client * const c,
+		      const struct mqtt_evt *evt)
+{
+	int err;
+
+	err = aws_fota_mqtt_evt_handler(c, evt);
+	if (err > 0) {
+		/* Event handled by FOTA library so we can skip it */
+		return;
+	}
+
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			printk("MQTT connect failed %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] MQTT client connected!\n", __func__, __LINE__);
+		if (err) {
+			printk("Unable to initialize AWS jobs upon connection\n");
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		printk("[%s:%d] MQTT client disconnected %d\n", __func__,
+		       __LINE__, evt->result);
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+
+		printk("[%s:%d] MQTT PUBLISH result=%d len=%d\n", __func__,
+		       __LINE__, evt->result, p->message.payload.len);
+		err = publish_get_payload(c, p->message.payload.len);
+		if (err) {
+			printk("mqtt_read_publish_payload: Failed! %d\n", err);
+			printk("Disconnecting MQTT client...\n");
+
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+
+		if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+			const struct mqtt_puback_param ack = {
+				.message_id = p->message_id
+			};
+
+			/* Send acknowledgment. */
+			err = mqtt_publish_qos1_ack(c, &ack);
+			if (err) {
+				printk("unable to ack\n");
+			}
+		}
+		data_print("Received: ", payload_buf, p->message.payload.len);
+		break;
+	}
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			printk("MQTT PUBACK error %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] PUBACK packet id: %u\n", __func__, __LINE__,
+		       evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			printk("MQTT SUBACK error %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] SUBACK packet id: %u\n", __func__, __LINE__,
+		       evt->param.suback.message_id);
+		break;
+
+	default:
+		printk("[%s:%d] default: %d\n", __func__, __LINE__,
+		       evt->type);
+		break;
+	}
+}
+
+
+/**@brief Resolves the configured hostname and
+ * initializes the MQTT broker structure
+ */
+static void broker_init(void)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo *addr;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM
+	};
+
+	err = getaddrinfo(CONFIG_AWS_HOSTNAME, NULL, &hints, &result);
```

Comment:
`result` will containt a list of `struct addrinfo` structures, and `addr` is used later to iterate overt that list.  You need to keep the head address though, so that you can release the memory later with `freeaddrinfo`.